[16.0][IMP] Take allow_out_payment into account on payment order

lmarion-source opened the pull request:
A previous MR existed but only when payment order was triggered from invoices.
We added the case also directly from confirm button on payment order

Original PR: https://github.com/OCA/bank-payment/pull/1177

Closes #1330

astirpe committed:
[16.0][IMP] account_payment_order: check that bank allows out payments

astirpe committed:
[IMP] account_payment_mode: set allow_out_payment on demo banks

acsonefho commented on account_payment_order/models/account_move.py:
```diff
@@ -151,6 +151,13 @@ def create_account_payment_line(self):
+            for line in applicable_lines.filtered(
+                lambda l: l.account_id.account_type == "liability_payable"
+            ):
+                line._check_bank_allows_out_payments()
```
As you already loop in your function, you can skip one level:
`applicable_lines.filtered(lambda l: l.account_id.account_type == "liability_payable")._check_bank_allows_out_payments()`

acsonefho reviewed (changes requested):
GJ :muscle: 
Just a minor change :pray:

adrienpeiffer commented:
/ocabot merge minor

pedrobaeza commented on account_payment_order/models/account_move.py:
```diff
@@ -151,6 +151,11 @@ def create_account_payment_line(self):
                         "order": payment_lines.order_id.mapped("name"),
                     }
                 )
+
```
Please don't introduce empty lines inside the method

```suggestion
```

pedrobaeza commented on account_payment_order/models/account_payment_order.py:
```diff
@@ -336,8 +336,10 @@ def draft2open(self):
+                    payline.move_line_id._check_bank_allows_out_payments()
                 except UserError as e:
                     payline_err_text.append(e.args[0])
+
```
No empty lines

```suggestion
```

pedrobaeza reviewed (changes requested):
I have stopped the merge. AFAIK, @sbidoul commented that wants to be able to add the lines, but not to confirm the order, and this is pre-filtering the non-allowed bank accounts previously if I understand the code right.

adrienpeiffer commented:
@pedrobaeza Not sure to understand. The check is done in draft2open method.

pedrobaeza commented:
I mean this code for example:

https://github.com/OCA/bank-payment/pull/1340/files#diff-bb6441f4ea06462fe1aae1f093f82c4e57460864b29b209676a6c7ce14e49ef9R156

It's avoiding to include the lines in the order.

adrienpeiffer commented:
Isn't that exactly the code approved in #1177 ?

pedrobaeza commented:
OK, I based my review in https://github.com/OCA/bank-payment/pull/1177#issuecomment-2304668141. Let me check again.

adrienpeiffer commented:
@lmarion-source Could you remove blank lines (cfr pedro's review)
@sbidoul If I undersood correctly, you prefer to be able to add lines from invoces even bank account is not validated ?
So basically, we have to remove this line https://github.com/OCA/bank-payment/pull/1340/files#diff-bb6441f4ea06462fe1aae1f093f82c4e57460864b29b209676a6c7ce14e49ef9R158

sbidoul commented:
> If I undersood correctly, you prefer to be able to add lines from invoces even bank account is not validated ?

It is my impression that Odoo Enterprise works like that, and it makes sense to me, since the process of preparing payment orders is distinct from the process of validating bank accounts, and often done by different persons. So I think it's better to let both processes run in parallel and block at the last moment. Cherry on the cake would be a warning when an non validated bank account is used in a payment line, but that can be done in another PR.

gdgellatly commented:
Hmm, we solved this completely differently. I get the minimal touch here, but we made a seperate module that hit all the places Odoo gets this wrong, mostly purchase order and account move but also payment with a single function on bank to get a valid out payment bank account. Then overrode `def _compute_available_partner_bank_ids(self):` which for me seems to be the place this belongs

sbidoul commented:
Reopening as I think it needs another review round after the last commits. What's the status of this?

Cedric-Pigeon reviewed (approved):
I like the approach of letting the user making the payment order and block at the end in case of unvalidated account.
I would be really interested to get it merge soon as I'd need to forward-port it to v17.0

Cedric-Pigeon commented:
@lmarion-source Could you please resolve conflicts ?

AnizR commented:
ping @lmarion-source

AnizR commented on account_payment_order/models/account_payment_order.py:
```diff
@@ -338,6 +338,12 @@ def draft2open(self):
             for payline in order.payment_line_ids:
                 try:
                     payline.draft2open_payment_line_check()
+                    if (
+                        self.env["ir.config_parameter"]
+                        .sudo()
+                        .get_param("account_payment_order.use_allow_out_payment")
+                    ):
```
```suggestion
                    if str2bool(
                        self.env["ir.config_parameter"]
                        .sudo()
                        .get_param("account_payment_order.use_allow_out_payment")
                    ):
```

AnizR reviewed (changes requested):
Almost good, thanks!

AnizR reviewed (approved):
Code LGTM

Cedric-Pigeon reviewed (approved):
LGTM

sbidoul commented:
@pedrobaeza can you update your review here?

sbidoul commented on account_payment_order/models/account_payment_order.py:
```diff
@@ -338,6 +339,12 @@ def draft2open(self):
+                        .sudo()
+                        .get_param("account_payment_order.use_allow_out_payment")
+                    ):
+                        payline._check_bank_allows_out_payments()
```
Don't we need to check that `payment_type` is `outbound` here? Because this check does not make sense for a debit order.

sbidoul replied:
@lmarion-source I looks like you have not handled this comment?

sbidoul commented on account_payment_order/models/account_move_line.py:
```diff
@@ -71,10 +71,10 @@ def _prepare_payment_line_vals(self, payment_order):
         if payment_order.payment_type == "outbound":
             amount_currency *= -1
-        partner_bank_id = self.partner_bank_id.id or first(self.partner_id.bank_ids).id
+        partner_bank_id = self.partner_bank_id or first(self.partner_id.bank_ids)
```
```suggestion
        partner_bank = self.partner_bank_id or first(self.partner_id.bank_ids)
```

pedrobaeza commented:
@lmarion-source can you please check the latest comments?

lmarion-source committed:
[16.0][IMP] Take allow_out_payment into account on payment order
A previous MR existed but only when payment order was triggered from invoices.
We added the case also directement from confirm button on payment order

original PR: https://github.com/OCA/bank-payment/pull/1177

sbidoul commented on account_payment_order/wizard/res_config_settings.py:
```diff
@@ -10,3 +10,7 @@ class ResConfigSettings(models.TransientModel):
         related="company_id.transfer_journal_id", readonly=False
     )
+    use_allow_out_payment = fields.Boolean(
+        string=" Use allow out payment on partner bank",
```
```suggestion
        string="Verify that Send Money is enabled on partner bank accounts",
```

Is this easier to understand by non technical users?

sbidoul commented on account_payment_order/wizard/res_config_settings.xml:
```diff
@@ -23,6 +23,19 @@
+                            </div>
+                        </div>
+                    </div>
+                </div>
```
This ends up in the next section. Not sure this is fixable in this PR, though.

<img width="550" height="612" alt="Image" src="https://github.com/user-attachments/assets/fa4123f3-3b60-4c4b-8f7a-761678bc36c3" />

sbidoul commented:
rebased

marcos-mendez reviewed (approved):
**Automated Review -- Tests Passed**

All tests for `account_payment_order` passed successfully on Odoo 16.0.
Environment: Minikube + K8s Job + oca-ci/py3.10-odoo16.0

---
### ⚠️ PR Aging Alert: CRITICAL

This PR by @lmarion-source has been **waiting for 561 days** — that is over **18 months** without being merged or closed.

> _Every ignored PR is a contributor who might not come back. Review time matters. ([OCA Aging Report](https://git.pop.coop/pop/oca-reviewer))_

---
### Reciprocal Review Request



Hi everyone! Thanks for your contribution to OCA. I reviewed and approved this PR. If any of you have a moment, I would really appreciate a review on my open PR(s):

**My open PRs across OCA:**
- [hr-attendance#262 [16.0][ADD] Hr_attendance_idsecure: iDSecure (ControliD) attendance integration](https://github.com/OCA/hr-attendance/pull/262)
- [stock-logistics-workflow#2276 [16.0][ADD] stock_move_line_devaluation](https://github.com/OCA/stock-logistics-workflow/pull/2276)
- [stock-logistics-workflow#2275 [16.0][ADD] Stock move line analytic account](https://github.com/OCA/stock-logistics-workflow/pull/2275)
- [stock-logistics-workflow#2268 [16.0][ADD] stock_move_line_picking_partner](https://github.com/OCA/stock-logistics-workflow/pull/2268)
- [purchase-workflow#2694 [16.0][IMP]Purchase workflow added to review state & exception fix](https://github.com/OCA/purchase-workflow/pull/2694)

Reviewing each other's work helps the whole community move forward faster. Thank you!

---
_Automated review by [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer) + qwen3-coder:30b_
